### ISSUE TITLE
Problematic interaction with implicit conversion gko::dim

### ISSUE THREAD

**Issue by upsj** [ginkgo-project/ginkgo#645]
I just ran into a really nasty usability issue which is something between an interface inconsistency for Identity and an effect of the implicit conversion gko::dim -> bool:

You write
```
gko::matrix::Csr<...>::create(exec, size);
```
to get a matrix of the given size. If you try the same for Identity
```
gko::matrix::Identity<...>::create(exec, size);
```
you get a 1x1 matrix instead. The reason for this is that Identity expects a size_type, not a gko::dim.
So since gko::dim is non-zero, it gets implicitly converted to bool (true) and then to size_type (1)

Is this something we would expect? To fix this, would it be better to remove the bool overload (which I would prefer, since this kind of implicit conversion from a 2D object to a 1D object should not happen) or change the interface to Identity?

**Comment by tcojean** [comment]
I would indeed drop the implicit conversion to bool. Is there an implicit conversion of `gko::dim<1>` or `gko::dim<2>` to size_type also? If not, I would go for added a new constructor which takes a `gko::dim`. At a later point (major release), we can remove the `size_type` interface altogether.

**Comment by thoasm** [comment]
I agree with @tcojean on adding this constructor as it is not interface breaking and would resolve that issue for now.
Removing the implicit conversion to bool, however, is interface breaking since I am pretty sure we use that quite extensively for checking if a matrix is non-empty:
```c++
if (mtx->get_size()) {
    // Code if matrix is non-empty / has a valid size
} else {
    // if it is empty / not valid
}
```

**Comment by tcojean** [comment]
That's a good point, I did not think about that use of dim. Adding an extra constructor though should remove the issue without having to remove the implicit bool conversion actually, so maybe that's enough of a fix.
